Add intro to dune chapter (#6)
melange-re/melange-for-react-devs: docs/intro-to-dune/index.md

- The `dune-project` file describes the metadata for your project, primarily:
  - The version of Dune you're using
  - The version of the Melange plugin for Dune you're using
  - The name of your project
- `dune` files describe things of interest to Dune, for example:
  - Which directories to include and which to exclude
  - Which directories contain code that should be transpiled to JavaScript,
    using the `melange.emit` stanza
- Make is a build automation tool heavily used in the OCaml world
  - `Makefile`s contain rules which can run commands based on the target you
    pass to the `make` command

## Solutions

<b>1.</b> Creating a separate app for Celsius Converter, with its own `dune`,
`Makefile`, and `Index.re` files, should look something like
[this](https://github.com/melange-re/melange-for-react-devs/tree/develop/src/celsius-converter-option).

<b>2.</b> Deleting `reason-react-ppx` from `src/counter/dune` will result in a
compilation error:

```
File "src/counter/Counter.re", line 5, characters 2-6:
5 |   <div
      ^^^^
Error: Unbound value div
```

That's because putting `reason-react-ppx` in the `preprocess/pps` field will

> transform function calls to `div` (which isn't defined anywhere) into calls to

Reviewer: Maybe we can link somewhere around here to the JSX section of the Reason docs, which explain how `<div/>` is converted: https://reasonml.github.io/docs/en/jsx.

Author: Added footnote with two links to more JSX info in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/0a6687827b600dd2a8dfa1c7eab0abb0c14a2199

# Introduction to Dune

Depending on how you've been following along, you may have several components in
your project. Since these components don't have much in common with each other,
it makes sense to put them in separate, independent apps. To do that, we'll have
to dive into [Dune](https://dune.build/), a build system designed for OCaml
projects.

> 

Reviewer 2: I miss the why on wanting to use dune:

- To organise your code properly and create different libraries
- To create a different arquitecture
- ...

Reviewer: Some features listed in https://melange.re/v2.1.0/build-system/#features.

Author: I added a sentence about why we use Dune a link to the features section in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/be21ece850e3c046f1a94b4a0f2d6d3726693508

## `dune-project` file

The `dune-project` specifies metadata for a project, and should appear in the
root directory of your project. If you've been using the [starter
project](/installation/#download-and-run-the-starter-project),
then you've been using Dune this whole time and therefore already have a
`dune-project` file:

<<< @/../dune-project{clj}

The line

```clj
(using melange 0.1)
```

is necessary because we have to manually enable the Melange plugin for Dune in
order to use it. Note that the version of the Melange Dune plugin is independent

> of the version of Melange we're using.

Reviewer: Dune documentation calls these `using` things "extensions":

https://dune.readthedocs.io/en/stable/melange.html#simple-project

```suggestion
is necessary because we have to manually enable the Melange extension for Dune in
order to use it. Note that the version of the Melange extension for Dune is independent
of the version of Melange we're using.
```

Author: Renamed plugin to extension in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/e6fa7475450429aeadc40366c39ff8dafd0fc351

> Technically, `dune-project` accepts many more metadata fields, but it’s best to

Reviewer: Can link to https://dune.readthedocs.io/en/stable/dune-files.html#dune-project:

```suggestion
Technically, `dune-project` accepts [many more metadata fields](https://dune.readthedocs.io/en/stable/dune-files.html#dune-project), but it’s best to
```

keep it minimal. Other than `name`, it makes more sense to put the rest of your
project's metadata fields in your `.opam` file, which we'll cover
[later](/todo/).

`dune-project` files use
[S-expressions](https://en.wikipedia.org/wiki/S-expression), which might make
you think of the [Lisp programming
language](https://en.wikipedia.org/wiki/Lisp_(programming_language)). However,
S-expressions are just a convenient syntax for encoding structured data, and
Dune doesn't have the power of a full scripting language.

Each S-expression at the top level is a known as a *stanza*. All the possible
stanzas you can use in `dune-project` can be found in Dune’s [Stanza

> Reference](https://dune.readthedocs.io/en/stable/dune-files.html#dune-project).

Reviewer: Ah I can see the docs are linked here, feel free to ignore my suggestion above.

## `dune` files

Besides `dune-project`, Dune also looks at the `dune` files in our project.
Basically, `dune` files tell Dune about directories, executables, libraries,
tests, and anything else of interest. For example, here's the `dune` file inside
the root directory of your project:

<<< @/../dune{clj}

Like `dune-project`, a `dune` file consists of one or more *stanzas*. The only
stanza here is
[dirs](https://dune.readthedocs.io/en/stable/dune-files.html#dirs), which tells
Dune which directories to include in the build. Note that the stanzas accepted
in `dune` files are not the same as the ones accepted by `dune-project`. See all
possible `dune` stanzas in Dune's [Stanza
Reference](https://dune.readthedocs.io/en/stable/dune-files.html#dune).

## `melange.emit` stanza

> Let's now take a look at the `src/dune` file:

Reviewer: Shouldn't this be `src/counter/dune`? That's the path used below.

Author: I agree it's a little confusing, but basically `src/dune` gets moved into `src/counter/dune` a little later in the chapter. Do you think I should add a comment (for authors, not readers) to clarify?

Reviewer: Ah, yes. A note for authors would help 👍

Author: Added comment explaining confusing reference to src/dune in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/50ea4e2b9bc03f8a7f201d49a45d7f8f91d783f9

Author: Actually, after https://github.com/melange-re/melange-for-react-devs/pull/6/commits/c0ae77d8a0c30e8461b99cc3cda65d40352db031, this comment is no longer needed because melange-for-react-devs-template's `dune` file no longer matches this repo's root `dune` file

<<< @/../src/counter/dune{clj}

The main stanza of interest for us is
[melange.emit](https://dune.readthedocs.io/en/stable/melange.html#melange-emit),
which tells Dune to turn our OCaml files into JavaScript files. The fields we
give to `melange.emit` here are `target`, `libraries`, `preprocess`, and
`module_systems`, which are ones that we need to use for pretty much every
Melange project.

A note about the `target` field: It is referenced in the `serve` npm script,
whose command is

```bash
webpack serve --open --mode development --entry ./_build/default/src/output/src/Index.js
```

If you change the value of `target` to, say, `stuff`, the value of the `--entry`
option must be changed to

```
./_build/default/src/stuff/src/Index.js

> ```

Reviewer: There is a section that explains these paths in the melange docs, maybe worth linking: https://melange.re/v2.1.0/build-system/#javascript-artifacts-layout

Author: Added a link to that section in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/e7c52e18e1a9f2c7ca1d191d0be4647c4e6170ae

browser, but if we also wanted to build JavaScript to run on Node, we'd include
another `melange.emit` stanza specifically for that. See
[melange-opam-template](https://github.com/melange-re/melange-opam-template/blob/main/src/dune)
for an example of building for Node.
:::

## Counter app directory

Create a new directory in `src` called `counter`, which will contain a new app
that only renders the [Counter component](/counter/). Then make sure this new
directory contains:

- `Counter.re` from `src`
- `dune` file from `src`
- `Index.re` to render the app to the DOM

  <<< @/../src/counter/Index.re
- `Makefile` to serve the app using webpack dev server

  <<< @/../src/counter/Makefile{make}

After you've added those files, your `src/counter` directory should look like
this:

```
src/counter
├─ Counter.re
├─ dune
├─ Index.re
└─ Makefile
```

## Why Makefile

You might be wondering why we're using a `Makefile` instead of adding another
npm script. In the world of OCaml, usage of [the Make build automation
tool](https://en.wikipedia.org/wiki/Make_(software)) is pretty much ubiquitous,
and this is not likely to change. So let's take a short detour to explain the

> `Makefile` we just added.

Reviewer 2: I would say that using Makefile is a personal preference, you can use anything really. 
 (justfile, even npm scripts if u want).

In fact, most people don't use Makefiles at all.

Reviewer: The main reason why we add Makefiles in Melange projects is for the creation of the opam switch and initial installation of opam packages, because that's the only command we can't safely write as a Dune alias, as the `dune` exe might not be available yet.

Otherwise, all other commands in the makefile are just pointing to dune commands:

https://github.com/melange-re/melange-for-react-devs/blob/f7ac1a701e0c98d215dc4ce6c1031b6ac77a01e0/Makefile#L32-L58

Author: Instead of saying Make usage is ubiquitous, I just say it's common. https://github.com/melange-re/melange-for-react-devs/pull/6/commits/5ca4d90edbfb7b79c7ce36279256065829dab015

## Anatomy of Makefile

In case you're not familiar with Make, here's a breakdown of
`src/counter/Makefile`:

<<< @/../src/counter/Makefile{make}

- There's only a single rule here, and its *target* is `serve`.
- `.PHONY: serve` is used to indicate that `serve` is a [phony
  target](https://makefiletutorial.com/#phony), meaning `serve` is not a file we

>   want to generate but just a label for an action. In a simple `Makefile` like

Reviewer: Not sure if it's worth establishing a parallel here between makefile phony targets and dune aliases, as they are similar.

Author: If I cover dune aliases in the book, I assume it will be much farther down the line, when I introduce a small OCaml server app. So I think I'll skip the reference for now.

Reviewer: Sounds good.

```

We basically copied the `serve` npm command from `package.json` to create the
recipe for this rule, but instead of a fixed value for the `--entry` option, we
use

```
./_build/default/src/$(app)/output/src/$(app)/Index.js
```

This means that the entry script served by webpack dev server depends on the
`app` environment variable, which is provided by `src/counter/Makefile`.

You're now ready to run the new Counter app you created! Go into the
`src/counter` directory and run

```
make serve
```

It should open a new tab in your default browser to <a
href="http://localhost:8080/" target="_blank" rel="noreferrer
noopener">http://localhost:8080/</a> and display your Counter component.

## Serving a default app

If you run `make serve` in the root directory, you'll get an error that looks
something like this:

```
ERROR in main
Module not found:
Error: Can't resolve './_build/default/src//output/src//Index.js' in '...'
```

That's because if `app` environment variable isn't provided, the value passed to
the `--entry` option is no longer a valid path. We can define a default value
for `app` by putting this near the top of `Makefile`:

```make
ifndef app
	app := counter
endif

> ```

Reviewer 2: is this necessary? Seems very long and not adding much to the setup.

Talking about the app env var and so on, the serving is great.

Author: Do you mean the "Serving a default app" section? 

I agree it's not essential, and could be deleted without losing much.

Author: Deleted 'Serving a default app' section in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/00be2116875700b99caa20daefcc2a813ef70892

- The `dune-project` file describes the metadata for your project, primarily:
  - The version of Dune you're using
  - The version of the Melange plugin for Dune you're using
  - The name of your project
- `dune` files describe things of interest to Dune, for example:
  - Which directories to include and which to exclude
  - Which directories contain code that should be transpiled to JavaScript,
    using the `melange.emit` stanza
- Make is a build automation tool heavily used in the OCaml world
  - `Makefile`s contain rules which can run commands based on the target you
    pass to the `make` command

## Solutions

<b>1.</b> Creating a separate app for Celsius Converter, with its own `dune`,
`Makefile`, and `Index.re` files, should look something like
[this](https://github.com/melange-re/melange-for-react-devs/tree/develop/src/celsius-converter-option).

<b>2.</b> Deleting `reason-react-ppx` from `src/counter/dune` will result in a
compilation error:

```
File "src/counter/Counter.re", line 5, characters 2-6:
5 |   <div
      ^^^^
Error: Unbound value div
```

That's because putting `reason-react-ppx` in the `preprocess/pps` field enables

> JSX in `.re` files.

Reviewer 2: That's not entirely correct. Here JSX is enabled (in all reason files is), but isn't treated for any ppx so the function `div` remains untouched and since isn't defined anywhere, it complains.

Author: I'm not sure I understand your explanation. Can I say something more like this?

```suggestion
That's because putting `reason-react-ppx` in the `preprocess/pps` field allows
all React element functions like `div` to be visible to your JSX code.
```

Reviewer: I think what Dave meant is that by default, Reason parser already processes JSX syntax, with or without `reason-react-ppx`.

For example, this reason code:

```reason
let t = <div />;
```

is equivalent to this other:

```reason
let t = [@JSX] div(~children=[], ());
```

You can see so by running refmt in the latter ([try in playground](https://melange.re/v2.1.0/playground/?language=Reason&code=bGV0IHQgPSBbQEpTWF0gZGl2KH5jaGlsZHJlbj1bXSwgKCkpOwo%3D&live=offis)).

What the `reason-react-ppx` does is to transform expressions annotated with `@JSX` attribute into calls to `React.createElement` or `ReactDOM.createElement`.

> That's because putting `reason-react-ppx` in the `preprocess/pps` field allows all React element functions like `div` to be visible to your JSX code.

I am not sure I understand what that ^ means. I would say something like:

> That's because putting `reason-react-ppx` in the `preprocess/pps` field will transform function calls to `div` into calls to `React.createElement("div", ...)`.

Author: Changed in https://github.com/melange-re/melange-for-react-devs/pull/6/commits/2e9aae9bc19c4ab66c98e5702fc2b40a928eb8bb to 

> That's because putting `reason-react-ppx` in the `preprocess/pps` field will
transform function calls to `div` (which isn't defined anywhere) into calls to
`React.createElement("div", ...)`.
